### PR TITLE
Add a basic mechanism for parsing server-sent events

### PR DESCRIPTION
**Pull request by JakeWharton**

### CODE REVIEW

**Comment by JakeWharton** [comment]
PR-ing to get initial parser and tests in. This should allow faster iteration on the consumption API (iterator-style to looper+callback) and making the Okio interaction more efficient in future PRs.

**Comment by yschimke** [comment]
Much better in the callback form since the server is sending them, asking client to consume is problematic.

File: okhttp-sse/src/test/java/okhttp3/internal/sse/Event.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.sse;
+
+import javax.annotation.Nullable;
+
+final class Event {
+  final @Nullable String id;
+  final @Nullable String type;
```
Comment (yschimke) [inline okhttp-sse/src/test/java/okhttp3/internal/sse/Event.java]: nit: name? why mix type and name here?

File: okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    void onMessage(@Nullable String id, @Nullable String event, String data);
+    void onRetryUpdated(long timeMs);
+  }
+
+  private final BufferedSource source;
+  private final Callback callback;
+
+  private String lastId = null;
+
+  public ServerSentEventReader(BufferedSource source, Callback callback) {
```
Comment (yschimke) [inline okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.java]: where is Callback?

File: okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+import java.io.IOException;
+import javax.annotation.Nullable;
+import okio.BufferedSource;
+import okio.ByteString;
+
+public final class ServerSentEventReader {
+  private static final ByteString CRLF = ByteString.encodeUtf8("\r\n");
+
+  public interface Callback {
```
Comment (yschimke) [inline okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.java]: Will you expand this to include open, closed, error, reconnect events?

File: okhttp-sse/src/test/java/okhttp3/internal/sse/Event.java (hunk 1 above)
Comment (JakeWharton) [reply okhttp-sse/src/test/java/okhttp3/internal/sse/Event.java]: Should be type everywhere

File: okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.java (hunk 3 above)
Comment (JakeWharton) [reply okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.java]: This is an internal callback like WebSocketReader's. It's only for reporting things that come across the stream. The user version of the callback will have those events though.

File: okhttp-sse/src/test/java/okhttp3/internal/sse/Event.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+  Event(@Nullable String id, @Nullable String type, String data) {
+    if (data == null) throw new NullPointerException("data == null");
+    this.id = id;
+    this.type = type;
+    this.data = data;
+  }
+
+  @Override public String toString() {
+    return "Event{id='" + id + "', name='" + type + "', data='" + data + "'}";
```
Comment (yschimke) [inline okhttp-sse/src/test/java/okhttp3/internal/sse/Event.java]: Should this be type=?

File: okhttp-sse/src/test/java/okhttp3/internal/sse/Event.java (hunk 1 above)
Comment (yschimke) [reply okhttp-sse/src/test/java/okhttp3/internal/sse/Event.java]: I this you missed the toString then.

File: okhttp-sse/src/test/java/okhttp3/internal/sse/Event.java (hunk 4 above)
Comment (JakeWharton) [reply okhttp-sse/src/test/java/okhttp3/internal/sse/Event.java]: Yes

Comment (JakeWharton) [reply okhttp-sse/src/test/java/okhttp3/internal/sse/Event.java]: https://github.com/square/okhttp/pull/4058

File: okhttp-sse/src/test/java/okhttp3/internal/sse/ServerSentEventIteratorTest.java [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+        + "\n");
+    assertEquals(new Event(null, null, "test"), callbacks.remove());
+    assertEquals(new Event(null, null, "test"), callbacks.remove());
+  }
+
+  @Test public void leadingWhitespace() throws IOException {
+    consumeEvents(""
+        + "data:  test\n"
+        + "\n");
+    assertEquals(new Event(null, null, " test"), callbacks.remove());
```
Comment (swankjesse) [inline okhttp-sse/src/test/java/okhttp3/internal/sse/ServerSentEventIteratorTest.java]: ick. this protocol is icky

Comment (swankjesse) [reply okhttp-sse/src/test/java/okhttp3/internal/sse/ServerSentEventIteratorTest.java]: If implementing a server you should explicitly add a space or risk losing it if the application provides one

File: okhttp-sse/src/test/java/okhttp3/internal/sse/ServerSentEventIteratorTest.java [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+  }
+
+  @Test public void idReusedAcrossEvents() throws IOException {
```
Comment (swankjesse) [inline okhttp-sse/src/test/java/okhttp3/internal/sse/ServerSentEventIteratorTest.java]: omg

Comment (swankjesse) [reply okhttp-sse/src/test/java/okhttp3/internal/sse/ServerSentEventIteratorTest.java]: this protocol does not work as I would have expected

File: okhttp-sse/src/test/java/okhttp3/internal/sse/ServerSentEventIteratorTest.java [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    consumeEvents(""
+        + "data: first event\n"
+        + "id: 1\n"
+        + "\n"
+        + "id: 2\n"
+        + "\n"
+        + "data: second event\n"
+        + "\n");
+    assertEquals(new Event("1", null, "first event"), callbacks.remove());
+    assertEquals(new Event("1", null, "second event"), callbacks.remove());
```
Comment (swankjesse) [inline okhttp-sse/src/test/java/okhttp3/internal/sse/ServerSentEventIteratorTest.java]: omg again

File: okhttp-sse/src/test/java/okhttp3/internal/sse/ServerSentEventIteratorTest.java [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+    assertEquals(new Event("1", null, "first event"), callbacks.remove());
+  }
+
+  @Test public void retryInvalidFormatIgnored() throws IOException {
+    consumeEvents(""
+        + "retry: 22\n"
+        + "\n"
+        + "retry: hey"
+        + "\n");
+    assertEquals(22L, callbacks.remove());
```
Comment (swankjesse) [inline okhttp-sse/src/test/java/okhttp3/internal/sse/ServerSentEventIteratorTest.java]: assert no more callbacks?

File: okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+      if (crlf == -1L) {
+        return false;
+      }
+
+      if (crlf == 0L) {
+        skipCrAndOrLf();
+
+        if (data != null) {
+          lastId = id;
+          callback.onEvent(id, type, data.toString());
```
[the remark this answers is not in the document]
Comment (swankjesse) [reply okhttp-sse/src/main/java/okhttp3/internal/sse/ServerSentEventReader.java]: now that I’ve read the tests, this is deliberate! yipes

File: okhttp-sse/src/test/java/okhttp3/internal/sse/ServerSentEventIteratorTest.java (hunk 8 above)
Comment (JakeWharton) [reply okhttp-sse/src/test/java/okhttp3/internal/sse/ServerSentEventIteratorTest.java]: There's an `@After` which does it

**Pull request merged**